src/modules/Session/model.ts

```diff
@@ -0,0 +1,108 @@
+import { inputObjectType, objectType } from 'nexus/dist';
+import { PrismaClient, Prisma } from '@prisma/client';
+import { randomUUID } from 'crypto';
+import { ApolloError } from 'apollo-server-core';
+import { token } from '../../helpers';
+
+export const Session = objectType({
+  name: 'Session',
+  definition(t) {
+    t.nonNull.string('id');
+    t.nonNull.date('createdAt');
+    t.nonNull.string('createdBy');
+    t.date('referenceExpiryDate');
+    t.nonNull.string('referenceTokenId');
+    t.nonNull.boolean('isUserCreated');
+    t.string('name');
+    t.date('revokedAt');
+  },
+});
+
+export const SessionCreate = inputObjectType({
+  name: 'SessionCreate',
+  definition(t) {
+    t.date('referenceExpiryDate');
+    t.nonNull.string('name');
+  },
+});
+
+export const SessionCreateOutput = objectType({
+  name: 'SessionCreateOutput',
+  definition(t) {
+    t.nonNull.field('session', { type: 'Session' });
+    t.nonNull.string('token');
+  },
+});
+
+async function newSession(
+  prisma: PrismaClient,
+  session: Prisma.SessionCreateInput,
+) {
+  return prisma.session.create({
+    data: session,
+  });
+}
+
+async function listSessions(prisma: PrismaClient, userId: string) {
+  return prisma.session.findMany({
+    where: {
+      createdBy: userId,
+    },
+  });
+}
+
+async function revoke(prisma: PrismaClient, sessionId: string, userId: string) {
+  try {
+    return await prisma.session.update({
+      where: {
+        createdBy_id: {
+          id: sessionId,
+          createdBy: userId,
+        },
+      },
+      data: {
+        revokedAt: new Date(),
+      },
+    });
+  } catch (e) {
+    throw new ApolloError('Failed to update session', 'SESSION_UPDATE_FAILED');
+  }
+}
+
+export async function generateTokenAndSession(
+  prisma: PrismaClient,
+  userId: string,
+  session: { referenceExpiryDate?: Date; name?: string },
+  isUserCreated: boolean = false,
+) {
+  const createId = randomUUID();
+  const createdToken = token.generate(createId, session.referenceExpiryDate);
+  const formattedToken = token.format(createdToken);
+  const createData = {
+    ...session,
+    id: createId,
+    referenceTokenId: formattedToken,
+    isUserCreated,
+    creator: {
+      connect: {
+        id: userId,
+      },
+    },
+  };
+  const createdSession = await newSession(prisma, createData);
+  return {
+    token: createdToken,
+    session: createdSession,
+  };
+}
+
+export function getSessionCrud(prisma: PrismaClient) {
+  return {
+    listSessions: async (userId: string) => listSessions(prisma, userId),
```

Comment:
Instead of redefining those functions and passing them prisma every time, please closely follow the pattern that was agreed and approved in `getUserCrud`: one function that returns object with all methods and shares prisma context

Reply:
each resolver for user has now 1 more line , but i don't think it's critical since it looks more like a middleware pipe then a set of particular actions.